src/webhook/webhook.service.ts

```diff
@@ -0,0 +1,99 @@
+import { HttpService } from '@nestjs/axios';
+import {
+  ForbiddenException,
+  Injectable,
+  NotFoundException,
+  UnauthorizedException,
+} from '@nestjs/common';
+import { CreateWebhookRequest, FetchedWebhook, WebhookEvent } from './dto';
+import { WebhookRepository } from './webhook.repository';
+import { Recipe } from '@prisma/client';
+import Cryptr from 'cryptr';
+
+@Injectable()
+export class WebhookService {
+  constructor(
+    private readonly httpService: HttpService,
+    private readonly webhookRepository: WebhookRepository,
+  ) {}
+  cryptr = new Cryptr(process.env.CRYPTR_SECRET_KEY);
+
+  async createWebhook(
+    userId: number,
+    webhookData: CreateWebhookRequest,
+  ): Promise<void> {
+    const userWebhooks = await this.webhookRepository.getAllWebhooksByUserId(
+      userId,
+    );
+    if (userWebhooks.length >= +process.env.WEBHOOK_LIMIT) {
+      throw new ForbiddenException();
+    }
+
+    if (webhookData.token) {
+      const encryptedString = this.cryptr.encrypt(webhookData.token);
+      webhookData.token = encryptedString;
+    }
+
+    this.webhookRepository.createWebhook(userId, webhookData);
+  }
+
+  async deleteWebhook(userId: number, webhookId: number) {
+    const webhook = await this.webhookRepository.getWebhookById(webhookId);
+    if (!webhook) {
+      throw new NotFoundException();
+    }
+    if (userId !== webhook.userId) {
+      throw new UnauthorizedException();
+    }
+    this.webhookRepository.deleteUserWebhookById(webhookId);
+  }
+
+  async getWebhooksByUserId(userId: number): Promise<FetchedWebhook[]> {
+    const webhooks = await this.webhookRepository.getAllWebhooksByUserId(
+      userId,
+    );
+    return webhooks.map(({ userId, token, ...rest }) => {
+      return rest;
+    });
+  }
+
+  sendToWebhook(url: string, data: Recipe, token?: string, attempt = 0): void {
+    let decryptedToken: string;
+    if (token) {
+      decryptedToken = this.cryptr.decrypt(token);
+    }
+    const headersRequest = {
+      'Content-Type': 'application/json',
+      Authorization: `Bearer ${token}`,
+    };
+    this.httpService
+      .post(url, data, {
+        headers: headersRequest,
+      })
+      .subscribe({
+        error: () => {
+          const maxWaitingTime = 86400 * 1000;
+          const nextTryInSec = Math.min(Math.pow(2, attempt + 1) * 1000);
+          if (nextTryInSec > maxWaitingTime) {
+            throw new NotFoundException({
+              message: 'Could not send data to provided URL',
+            });
+          }
+          setTimeout(() => {
+            this.sendToWebhook(url, data, token, attempt + 1);
+          }, nextTryInSec);
```

Comment:
hmm, using `setTimeout` in this case might to be the best. Once there are more users in the system using webhooks, you will end up with a lot of `setTimeout` calls running in the memory that could affect backend performance. Also once you restart the server (e.g. because you are deploying a new version or an unexpected error occurred) you will lose all timeouts you have created.

I would change the approach here - use `cron` scheduler. You can run it e.g. every 1 minute and check if there are any webhook events that could be resent (only failed webhook events, new webhook events should be sent immediately, outside of cron).

Although using cron introduces a horizontal scaling problem (we want to run cron once - with 2 or more backend instances cron will be called once on each instance), we can fix it by either using semaphores (e.g. using Redlock) or moving cron logic to a separate micro service that will always have 1 instance only. For now - we don't have to care about this problem.